api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -227,132 +360,200 @@ private void maybeAlertOnCostThresholds(
    *
-   * <p>Note: return value may be null, to enable direct assignment to the nullable Profile field
+   * <p>1. The workspace is active 2. The workspace is deleted and either of the following is true
+   * 2.1. Free Tier Usage last updated time is null. This means that it wasn't calculated before
+   * 2.2. Workspace last updated time is null. This means we don't have enough info about the
+   * workspace, so we need to get its cost 2.3. Free Tier Usage time is before the Workspace last
+   * updated time. This means that the workspace got changed some time after our last calculation,
+   * so we need to recalculate it 2.4. Free Tier Usage time is after the Workspace last updated
+   * time, but the difference is smaller than a certain value. This case to account for charges that
+   * may occur after the workspace gets deleted and after the last cron had run.
    *
-   * @param user the user as represented in our database
-   * @return the total USD amount spent in workspaces created by this user, represented as a double
+   * @param allCostsInDbForUsers
+   * @return a {@link java.util.List} of the workspaces that require updates only
    */
-  @Nullable
-  public Double getCachedFreeTierUsage(DbUser user) {
-    return workspaceFreeTierUsageDao.totalCostByUser(user);
+  @NotNull
+  private List<WorkspaceCostView> findActiveOrDeletedWorkspacesThatRequireUpdates(
+      List<WorkspaceCostView> allCostsInDbForUsers) {
+
+    List<WorkspaceCostView> workspacesThatRequireUpdate =
+        allCostsInDbForUsers.stream()
+            .filter(
+                c ->
+                    (c.getActiveStatus() == 0)
+                        || (c.getActiveStatus() == 1
+                            && (c.getFreeTierLastUpdated() == null
+                                || c.getWorkspaceLastUpdated() == null
+                                || c.getFreeTierLastUpdated().before(c.getWorkspaceLastUpdated())
+                                || (c.getFreeTierLastUpdated().after(c.getWorkspaceLastUpdated())
+                                    && c.getFreeTierLastUpdated().getTime()
+                                            - c.getWorkspaceLastUpdated().getTime()
+                                        < getNumberOfDaysToConsiderForFreeTierUsageUpdate()))))
+            .collect(Collectors.toList());
+
+    logger.info(
+        String.format("Workspaces that require update %d", workspacesThatRequireUpdate.size()));
+
+    return workspacesThatRequireUpdate;
   }
 
   /**
-   * Does this user have remaining free tier credits? Compare the user-specific free tier limit (may
-   * be the system default) to the amount they have used.
+   * Requesting live costs from BQ in batches then updating it in a transactional method.
    *
-   * @param user the user as represented in our database
-   * @return whether the user has remaining credits
+   * @param allCostsInDbForUsers
+   * @param workspacesThatRequireUpdate
```

Comment:
If use param, suggest to add more meaning full explanation, instead of just var name..
Also this method does worth 1~2 line javadoc to explain.